File: KLine-Chart-OC/KLine-Chart-OC/NetWork/HTTPTool.m
//
//  HTTPTool.m
//  KLine-Chart-OC
//
//  Created by 何俊松 on 2020/3/11.
//  Copyright © 2020 hjs. All rights reserved.
//

#import "HTTPTool.h"

@interface HTTPTool()
@property(nonatomic,strong) NSURLSessionDataTask *currentDataTask;

@end

static HTTPTool *tool = nil;
@implementation HTTPTool

+(instancetype)tool {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        tool = [[self alloc] init];
    });
    return tool;
}


-(void)getData:(NSString *)period  complation:(void(^)(NSArray<KLineModel *> *models))complationBlock  {
    [_currentDataTask cancel];
    NSString *urls = [NSString stringWithFormat:@"%@%@%@",@"https://api.huobi.pro/market/history/kline?period=",period,@"&size=300&symbol=btcusdt"];
    NSURL *url = [[NSURL alloc] initWithString:urls];
    
    NSURLRequest *requst = [[NSURLRequest alloc] initWithURL:url];
    
    NSURLSession *session = [NSURLSession sharedSession];
    
    NSArray *datas = [self getLocalData];
    dispatch_async(dispatch_get_main_queue(), ^{
        complationBlock(datas);
    });
//    NSURLSessionDataTask *dataTask = [session dataTaskWithRequest:requst completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
//        if(error == nil) {
//           NSDictionary *dict =  [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingFragmentsAllowed error:nil];
//            NSArray<NSDictionary *> *dicts = dict[@"data"];
//            NSMutableArray *array = [NSMutableArray arrayWithCapacity:100];
//            for (int i = 0; i < dicts.count; i++) {
//                NSDictionary *item = dicts[i];
//                [array addObject:[[KLineModel alloc] initWithDict:item]];
//            }
//            dispatch_async(dispatch_get_main_queue(), ^{
//                complationBlock(array);
//            });
//        } else {
//            NSArray *datas = [self getLocalData];
//            dispatch_async(dispatch_get_main_queue(), ^{
//                complationBlock(datas);
//            });
//        }
//    }];
//    [dataTask resume];
}

-(NSArray<KLineModel *> *)getLocalData {
   NSString *path = [[NSBundle mainBundle] pathForResource:@"kline" ofType:@"json"];
   NSDate *data = [[NSData alloc] initWithContentsOfURL: [[NSURL alloc] initFileURLWithPath:path]];
    NSArray<NSDictionary *> *dicts = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingFragmentsAllowed error:nil];
//   NSArray<NSDictionary *> *dicts = dict[@"data"];
   NSMutableArray *array = [NSMutableArray arrayWithCapacity:100];
   for (int i = 10; i < 11; i++) {
       NSDictionary *item = dicts[i];
       [array addObject:[[KLineModel alloc] initWithDict:item]];
    }
    return array;
}



@end

File: KLine-Chart/KLine-Chart/NetWork/HTTPTool.swift
//
//  HTTPTool.swift
//  KLine-Chart
//
//  Created by 何俊松 on 2020/3/4.
//  Copyright © 2020 hjs. All rights reserved.
//

import UIKit

class HTTPTool: NSObject {
    static let tool = HTTPTool()

    var currentDataTask: URLSessionDataTask?

    func getData(period: String, complationBlock: @escaping (([KLineModel]) -> Void)) {
        currentDataTask?.cancel()
        let url = URL(string: "https://api.huobi.pro/market/history/kline?period=\(period)&size=2000&symbol=btcusdt")

        let requst = URLRequest(url: url!)

        let session: URLSession = URLSession.shared

        let dataTask: URLSessionDataTask = session.dataTask(with: requst) { data, _, error in
            if error == nil, let _data = data {
                guard let dict = try? JSONSerialization.jsonObject(with: _data, options: JSONSerialization.ReadingOptions.allowFragments) as? [String: Any] else {
                    fatalError()
                }
                print(dict)
                if let dicts = dict["data"] as? [[String: Any]] {
                    let datas = dicts.map { (dict) -> KLineModel in
                        KLineModel(dict: dict)
                    }
                    var addDatas: [KLineModel] = []
                    var newDatas: [KLineModel] = []
                    for idx in 0 ..< datas.count {
                        if idx < 100 {
                            addDatas.append(datas[idx])
                        } else {
                            newDatas.append(datas[idx])
                        }
                    }
                    KLineStateManger.manager.addDatas = addDatas
                    DispatchQueue.main.async {
                        complationBlock(newDatas)
                    }
                    return
                }
            }
            let datas = self.getLocalData()
            DispatchQueue.main.async {
                complationBlock(datas)
            }
        }
        currentDataTask = dataTask
        dataTask.resume()
    }

    func getLocalData() -> [KLineModel] {
        guard let path = Bundle.main.path(forResource: "kline", ofType: "json") else {
            fatalError()
        }
        guard let data = try? Data(contentsOf: URL(fileURLWithPath: path)) else {
            fatalError()
        }
        guard let dicts = try? JSONSerialization.jsonObject(with: data, options: JSONSerialization.ReadingOptions.allowFragments) as? [[String: Any]] else {
            fatalError()
        }
        let datas = dicts.map { (dict) -> KLineModel in
            KLineModel(dict: dict)
        }
        return datas
    }
}
